File: Python/src/leetcode_002.py
from typing import Optional

# Definition for singly-linked list.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    # This is needed for assert actual == expected
    def __eq__(self, other) -> bool:
        while self and other:
            if self.val != other.val:
                return False
            self = self.next
            other = other.next

        return self is None and other is None


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        p = dummy
        carry = 0

        while l1 or l2:
            x = 0
            y = 0

            if l1:
                x = l1.val
                l1 = l1.next

            if l2:
                y = l2.val
                l2 = l2.next

            s = x + y + carry
            carry = s // 10
            s = s % 10
            p.next = ListNode(s)
            p = p.next

        if carry:
            p.next = ListNode(carry)

        return dummy.next

File: C/src/leetcode_002.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
  int val;
  struct ListNode *next;
};

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) {
  struct ListNode *result = NULL;
  struct ListNode *current = NULL;
  struct ListNode *previous = NULL;
  int carry = 0;
  int sum = 0;

  while (l1 != NULL || l2 != NULL) {
    sum = carry;
    if (l1 != NULL) {
      sum += l1->val;
      l1 = l1->next;
    }
    if (l2 != NULL) {
      sum += l2->val;
      l2 = l2->next;
    }
    carry = sum / 10;
    current = malloc(sizeof(struct ListNode));
    current->val = sum % 10;
    current->next = NULL;
    if (result == NULL) {
      result = current;
    } else {
      previous->next = current;
    }
    previous = current;
  }
  if (carry > 0) {
    current = malloc(sizeof(struct ListNode));
    current->val = carry;
    current->next = NULL;
    previous->next = current;
  }
  return result;
}
